xmtp_mls/src/verified_key_package_v2.rs

```diff
@@ -0,0 +1,76 @@
+use openmls::{
+    credentials::{errors::BasicCredentialError, BasicCredential},
+    prelude::{
+        tls_codec::{Deserialize, Error as TlsCodecError},
+        KeyPackage, KeyPackageIn, KeyPackageVerifyError,
+    },
+};
+use openmls_rust_crypto::RustCrypto;
+use prost::{DecodeError, Message};
+use thiserror::Error;
+
+use crate::configuration::MLS_PROTOCOL_VERSION;
+use xmtp_proto::xmtp::identity::MlsCredential;
+
+#[derive(Debug, Error)]
+pub enum KeyPackageVerificationError {
+    #[error("TLS Codec error: {0}")]
+    TlsError(#[from] TlsCodecError),
+    #[error("mls validation: {0}")]
+    MlsValidation(#[from] KeyPackageVerifyError),
+    #[error("invalid lifetime")]
+    InvalidLifetime,
+    #[error("wrong credential type")]
+    WrongCredentialType(#[from] BasicCredentialError),
+    #[error(transparent)]
+    Decode(#[from] DecodeError),
+}
+
+pub struct VerifiedKeyPackageV2 {
```

Comment:
The old `VerifiedKeyPackage` actually checked the signatures in the credential, so you knew you could trust it.

I wonder if we want to either

A) Rename this so that it's clear users still need to check the identity update history for the inbox id to make sure it's valid
B) Add those checks in here and take a reference to the client to access the DB/API

Not a blocker. Just food for thought.

Reply:
hm,

That's a good catch. I think we should probably go with B), I think A might still lead to users glossing over that. If users _dont_ have a use for a key package that's actually invalid w.r.t InboxId, I think we should definitely include this check, despite added complexity with a Client

Reply:
Maybe it's a bit of both A and B.

This is kinda a half-verified key package. The MLS checks have passed, but not our Inbox ID authentication. That's still valuable for places like the MLS Validation Service, where you need to parse the key package and then later verify the inbox ID stuff.

So, maybe we have some sort of `ParsedKeyPackage` and `VerifiedKeyPackage` or whatever you want to name it. Then you can move from raw -> parsed -> verified, or stop in the middle.